src/bootstrap/min_config.rs

```diff
@@ -0,0 +1,430 @@
+use core::fmt;
+use std::{
+    cmp,
+    collections::HashMap,
+    env, fs,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+use serde_derive::Deserialize;
+
+use crate::{
+    cache::{Interned, INTERNER},
+    config::{self, set},
+    flags::Flags,
+    t,
+    util::output,
+};
+
+/// The bare minimum config, suitable for `bootstrap-shim`, but sharing code with the main `bootstrap` binary.
+/// We only want to include indispensable parts of configuration to have highest possible BC for bootstrap-shim.
```

Comment:
Can we avoid "BC" here? I assume that means "build caching", but it does not feel like a standard term to me.

Reply:
It means "backward compatibility". Pretty common and popular term. I will explicitly update it as "backward compatibility".

Reply:
Never seen it before, but I think in that case we need more here - what are the implications if we break compatibility? Shouldn't it be the same across both configs?

Reply:
I see, will provide more detailed informations

Reply:
I was quite wrong about backwards compatibility. Just realised that.

https://github.com/rust-lang/rust/blob/d3039c7671c615a7f822279a19c0205b2a05bf7b/src/bootstrap/min_config.rs#L20-L24

Is this clear enough?

Reply:
This is very much focusing on the literal usage of the struct, rather than why we have the separation and what reviewers/authors of changes should think through. If we just used Config, what would go wrong? I really want us to document that in the code, especially since it seems like we can't get a clear answer when asking the question now :) @jyn514 -- maybe you recall the details here?

My recollection is that we were concerned about the shim being out of date and not parsing config.toml as a result (i.e., a new field in config.toml would fail to parse with a slightly out of date shim). Is that the motivation here? It feels a little weak to me (we should be able to check that the shim people will be using is good enough for changes to config in CI, so it would just add some delay when adding and using new fields, I think?). But regardless, that should be clearly documented.

And, it seems like if we *are* worried about this kind of scenario, we should probably add a checkbox to https://github.com/rust-lang/rust/issues/94829 for validating it in CI.

But maybe I'm misremembering and we had some other scenario in mind.

Reply:
>  If we just used Config, what would go wrong?

I don't think things can go wrong, but it's unnecessary to parse whole config when you only need few of them for `bootstrap-shim` initialization.

I leave this to @jyn514 to give further details

Reply:
Yes, I think we're missing something. We should not pay all this complexity for a few milliseconds less CPU time, if that.

Reply:
Like you said, the performance wasn't really a consideration. There are two reasons I wanted to avoid parsing all of `Config`:
1. Unlike bootstrap itself, bootstrap-shim needs to be compatible across multiple versions of the Rust repo. I thought separating the fields it reads into another struct and adding a big "DO NOT CHANGE THIS LOGIC WITHOUT CHANGING THE SHELL SCRIPTS" comment would make it easier to maintain without accidentally regressing the shim.
2. When I wrote this originally, I wanted to make `bootstrap-shim` entirely separate from `bootstrap` itself except for the config.toml fields it reads, and maybe some things in `util.rs` and `build_helper`, with the goal of making it extremely fast to compile and download. Right now we're effectively downloading bootstrap twice, which is *fine*, but seems kind of wasteful. It ended up being a *lot* more work than I expected to separate them, but my hope was by at least separating the configs it will be easier to change the shim to not depend on all 90 of bootstrap's dependencies in the future.

If it's turning out to be a hassle, I don't mind using the same Config for both for now, but I think we should at least add the "DON'T TOUCH THESE FIELDS" comment and open a follow-up issue for making the shim smaller in the future.

Honestly if we're using the same config for both I'm not sure it helps to have a shim? at that point we may as well package bootstrap itself in each nightly, since they're effectively the same binary.

Reply:
> Unlike bootstrap itself, bootstrap-shim needs to be compatible across multiple versions of the Rust repo. I thought separating the fields it reads into another struct and adding a big "DO NOT CHANGE THIS LOGIC WITHOUT CHANGING THE SHELL SCRIPTS" comment would make it easier to maintain without accidentally regressing the shim.

I think this is a good start to such a comment -- but I think we also should dig a little deeper in the comment. In particular, I think this "must be compatible" piece is a *distinct* requirement from the one we typically have -- usually we always are checking the current commit and searching back to find the latest change to some tool and then downloading that.

But, the plan for bootstrap-shim is, I guess, different : we will be using a *global* install of it? I forget the details now, but this seems like *exactly* the content I would want in a comment on the minimal config so that we can read that and understand what the compatibility story should be. I'm also not sure how the shell scripts play into this yet, but in any case good to reflect into the comment I'm looking for here.

I don't have a huge stake in what we ship or don't ship, but it seems to me that at least in theory if the shim is a separate binary we *should* see LLVM/linkers cut out any unused dependencies. On the other hand, maybe that doesn't give us that much - seems like the shim ends up at least pulling in decompression (xz + gzip?), tar, etc, so it probably won't be tiny.

Anyway to summarize it seems to me that probably the next step is for @jyn514 to maybe either point to a hackmd (I seem to recall one) or otherwise reflect some of these thoughts into the comment, so we can move forward based on that. I think there's also some other comments on this PR I've left that seem to still be warranted to resolve from my perspective, but it's possible that with a more detailed description of the purpose/rationale/strategy for the shim and minimal config I'll change my mind :)

Sounds like we're making definite progress though.

Reply:
Here's the hackmd you're remembering: https://hackmd.io/@jynelson/S1Aa4cozo/edit

Reply:
once the blocker PR is merged, I will re-update the doc-comments by the given updates from @jyn514 

> Here's the hackmd you're remembering: https://hackmd.io/@jynelson/S1Aa4cozo/edit

FYI the link is broken for me. Returns 404 Not Found.

Reply:
fixed, thank you

Reply:
https://github.com/rust-lang/rust/blob/8ad89068e2558940b4184d46763ff9c6b325883b/src/bootstrap/min_config.rs#L20-L26

I didn't want to leave it as `to be filled`, because devs(who needs to make a change on config) might still be confused until we make the shim default. So I give the idea why it shouldn't be changed directly, and left a `FIXME` to update it once shim is default.

What do you think about the current change? @Mark-Simulacrum